### PR TITLE
Do not use namespaceSelector for checking individual feature state

### LINKED ISSUE #2945: featuregate api failing for non administrator user
**Bug description**
For TKGS non-administrator users, the CLI command ` tanzu cluster list -n ns01` is failing with the below error:
```
Error: error while checking feature 'vmware-system-tkg-clusterclass' status in namespace 'vmware-system-tkg': failed to get namespaces from NamespaceSelector: namespaces is forbidden: User "janedoe@example.com" cannot list resource "namespaces" in API group "" at the cluster scope
Error: exit status 1
```

Here are steps to reproduce:

- In the TKGS testbed, create a namespace (ns01) and a new user created(user1) and assign Edit permissions to namespace (ns01).
- Generate kubeconfig file by connecting to the TKGS cluster with kubectl vsphere login with the newly created user (user1).
- In my local system, CLI is installed, and tkgs supervisor is added as a new server (with `tanzu login` command).
- the command 'tanzu cluster list -n ns01' fails with above error

**Affected product area (please put an X in all that apply)**

- [ ] APIs
- [ ] Addons
- [x] CLI
- [ ] Docs
- [ ] IAM
- [ ] Installation
- [ ] Plugin
- [ ] Security
- [ ] Test and Release
- [ ] User Experience

**Expected behavior**
`tanzu cluster list `command should work
**Steps to reproduce the bug**
listed in description
**Version** (include the SHA if the version is not obvious)
➜  cli tanzu cluster version
v0.25.0-dev
**Environment where the bug was observed (cloud, OS, etc)**
Mac
**Relevant Debug Output (Logs, manifests, etc)**

### PR DESCRIPTION
**Pull request by rajathagasthya**
### What this PR does / why we need it

Using `namespaceSelector` to check activation state of an individual
feature in `FeatureActivatedInNamespace` function is not really
necessary as it adds extra RBAC requirements for the client to get/list
namespaces.

Removing it does not functionally change the behavior. The function will
still return false if the namespace doesn't exist because it wouldn't be
present in any feature gate's `status.namespaces` during the comparison.
At worst, this results in extra comparisons, but does not change the
result.

### Which issue(s) this PR fixes

Fixes #2945 

### Describe testing done for PR

WIP

### Release note

```release-note
NONE
```

### PR Checklist

- [x] Squash the commits into one or a small number of logical commits
      
- [x] Use good commit [messages](https://github.com/vmware-tanzu/tanzu-framework/blob/main/CONTRIBUTING.md)
- [x] Ensure PR contains terms all contributors can understand and links all contributors can access

### CODE REVIEW

**Review by danniel1205 — APPROVED** [review]
lgtm. thanks for the PR

**Comment by danniel1205** [comment]
adding `do-not-merge` label to wait for all the CI checks.

File: pkg/v1/sdk/features/featuregate/featuregate.go [hunk 1]
```diff
@@ -83,23 +83,23 @@ func ComputeFeatureStates(featureGateSpec configv1alpha1.FeatureGateSpec, featur
...
-		MatchExpressions: []metav1.LabelSelectorRequirement{
-			{Key: "kubernetes.io/metadata.name", Operator: metav1.LabelSelectorOpIn, Values: []string{namespace}},
-		},
-	}
-	return FeaturesActivatedInNamespacesMatchingSelector(ctx, c, selector, []string{feature})
+	return featuresActivated(ctx, c, []string{namespace}, []string{feature})
 }
 
 // FeaturesActivatedInNamespacesMatchingSelector returns true only if all the features specified are activated in every namespace matched by the selector.
+// Note the client passed in must have appropriate RBAC to list namespaces in order to apply the selector.
```
Comment (yharish991) [inline pkg/v1/sdk/features/featuregate/featuregate.go:92]: can we remove this func as it is not being used anymore?

**Review by yharish991 — APPROVED** [review]

File: pkg/v1/sdk/features/featuregate/featuregate.go [hunk 2]
```diff
@@ -83,23 +83,23 @@ func ComputeFeatureStates(featureGateSpec configv1alpha1.FeatureGateSpec, featur
...
 func FeaturesActivatedInNamespacesMatchingSelector(ctx context.Context, c client.Client, namespaceSelector metav1.LabelSelector, features []string) (bool, error) {
 	namespaces, err := featureutil.NamespacesMatchingSelector(ctx, c, &namespaceSelector)
 	if err != nil {
 		return false, err
 	}
+	return featuresActivated(ctx, c, namespaces, features)
+}
 
+// featuresActivated returns true only if all the features specified are activated in each input namespace.
+func featuresActivated(ctx context.Context, c client.Client, namespaces, features []string) (bool, error) {
```
Comment (sathyanarays) [inline pkg/v1/sdk/features/featuregate/featuregate.go:102]: Is it a valid scenario if two different feature gates have at least one namespace in common? If yes, this function may not work as expected. Please refer the following test:

The test should pass; but it fails.
```
func TestFeatureActivatedInNamespaceUtil(t *testing.T) {
	scheme, err := configv1alpha1.SchemeBuilder.Build()
	if err != nil {
		t.Fatal(err)
	}
	if err := k8sscheme.AddToScheme(scheme); err != nil {
		t.Fatal(err)
	}

	newNamespace := func(name string) *corev1.Namespace {
		return &corev1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: name, Labels: map[string]string{"kubernetes.io/metadata.name": name}}}
	}

	newFeatureGate := func(name string, namespaces, activated, deactivated, unavailable []string) *configv1alpha1.FeatureGate {
		return &configv1alpha1.FeatureGate{
			ObjectMeta: metav1.ObjectMeta{
				Name: name,
			},
			Status: configv1alpha1.FeatureGateStatus{
				Namespaces:          namespaces,
				ActivatedFeatures:   activated,
				DeactivatedFeatures: deactivated,
				UnavailableFeatures: unavailable,
			}}
	}

	testCases := []struct {
		description     string
		existingObjects []runtime.Object
		namespace       []string
		feature         []string
		want            bool
		err             string
	}{
		{
			description: "Namespace exists and is gated and feature is activated",
			existingObjects: []runtime.Object{
				newNamespace("tkg-system"),
				newFeatureGate("tkg-one", []string{"tkg-system"}, []string{"one"}, []string{"three", "four"}, []string{"five", "six"}),
				newFeatureGate("tkg-two", []string{"tkg-system"}, []string{"two"}, []string{"three", "four"}, []string{"five", "six"}),
			},
			namespace: []string{"tkg-system"},
			feature:   []string{"one", "two"},
			want:      true,
			err:       "",
		},
	}

	for _, tc := range testCases {
		t.Run(tc.description, func(t *testing.T) {
			fakeClient := fake.NewClientBuilder().WithScheme(scheme).WithRuntimeObjects(tc.existingObjects...).Build()
			got, err := featuresActivated(context.Background(), fakeClient, tc.namespace, tc.feature)
			if err != nil {
				if tc.err == "" {
					t.Errorf("no error string specified, but got error: %v", err)
				}
				if !strings.Contains(err.Error(), tc.err) {
					t.Errorf("error string=%q doesn't match partial=%q", tc.err, err)
				}
			} else if tc.err != "" {
				t.Errorf("error string=%q specified but error not found", tc.err)
			}
			if got != tc.want {
				t.Errorf("feature activation: got %t, want %t", got, tc.want)
			}
		})
	}
}
```

Comment (rajathagasthya) [reply pkg/v1/sdk/features/featuregate/featuregate.go:102]: > Is it a valid scenario if two different feature gates have at least one namespace in common?

Great question! This is not a valid scenario. We validate this in the webhook: https://github.com/vmware-tanzu/tanzu-framework/blob/73f58999601aec1660877fd86f470ecba3ede267/apis/config/v1alpha1/featuregate_webhook.go#L118-L119

**Pull request closed without merging**